framework/src/media/HttpInputDataSource.cpp

```diff
@@ -0,0 +1,292 @@
+#define CONFIG_HTTPSOURCE_DOWNLOAD_BUFFER_SIZE 4096
+#endif
+
+#ifndef CONFIG_HTTPSOURCE_DOWNLOAD_BUFFER_THRESHOLD
+#define CONFIG_HTTPSOURCE_DOWNLOAD_BUFFER_THRESHOLD 2048
+#endif
+
+namespace media {
+namespace stream {
+
+// Content-Type for audio stream
+static const std::string TAG_CONTENT_TYPE = "Content-Type:";
+static const std::string AAC_MIME_TYPE = "audio/aac";
+static const std::string AACP_MIME_TYPE = "audio/aacp";
+static const std::string MPEG_MIME_TYPE = "audio/mpeg";
+static const std::string OPUS_MIME_TYPE = "audio/opus";
+
+static const std::chrono::seconds WAIT_HEADER_TIMEOUT = std::chrono::seconds(3);
+static const std::chrono::seconds WAIT_DATA_TIMEOUT = std::chrono::seconds(3);
+
+HttpInputDataSource::HttpInputDataSource(const std::string &url)
+	: InputDataSource(), mUrl(url)
+{
+	medvdbg("url: %s\n", mUrl.c_str());
+}
+
+HttpInputDataSource::HttpInputDataSource(const HttpInputDataSource &source)
+	: InputDataSource(source), mUrl(source.mUrl)
+{
+}
+
+HttpInputDataSource &HttpInputDataSource::operator=(const HttpInputDataSource &source)
+{
+	InputDataSource::operator=(source);
+	return *this;
+}
+
+bool HttpInputDataSource::open()
+{
+	medvdbg("HttpInputDataSource::open!\n");
+
+	if (isPrepare()) {
+		medvdbg("HttpInputDataSource is already opened!\n");
+		return true;
+	}
+
+	if (mStreamBuffer == nullptr) {
+		mStreamBuffer = StreamBuffer::Builder()
+								.setBufferSize(CONFIG_HTTPSOURCE_DOWNLOAD_BUFFER_SIZE)
+								.setThreshold(CONFIG_HTTPSOURCE_DOWNLOAD_BUFFER_THRESHOLD)
+								.build();
+
+		if (mStreamBuffer == nullptr) {
+			meddbg("mStreamBuffer is nullptr!\n");
+			return false;
+		}
+
+		mStreamBuffer->setObserver(this);
+		mBufferReader = std::make_shared<StreamBufferReader>(mStreamBuffer);
+		mBufferWriter = std::make_shared<StreamBufferWriter>(mStreamBuffer);
+	}
+
+	if (mHttpStream == nullptr) {
+		mHttpStream = HttpStream::create();
+		if (mHttpStream == nullptr) {
+			meddbg("mHttpStream is nullptr!\n");
+			return false;
+		}
+	}
+
+	std::unique_lock<std::mutex> lock(mMutex);
+	mIsHeaderReceived = false;
+	mIsDataReceived = false;
+
+	mThread = std::thread([this]() {
+		medvdbg("download thread enter!\n");
+		//mHttpStream->addHeader("Icy-MetaData:1"); // not support now
+		mHttpStream->setHeaderCallback(HeaderCallback, (void *)this);
+		mHttpStream->setWriteCallback(WriteCallback, (void *)this);
+		if (!mHttpStream->download(mUrl)) {
+			medwdbg("download failed or terminated!\n");
+			// TODO: send network error code to upper layer later
+		}
+
+		mBufferWriter->setEndOfStream();
+		medvdbg("download thread exit!\n");
+	});
+
+	// wait for Content-Type header
+	if (!mCondv.wait_for(lock, WAIT_HEADER_TIMEOUT, [=]{ return mIsHeaderReceived; })) {
+		medwdbg("download:: wait header timeout!\n");
+		mBufferWriter->setEndOfStream();
+		return false;
+	}
+
+	audio_type_t audioType = AUDIO_TYPE_UNKNOWN;
+	if ((mContentType.find(AAC_MIME_TYPE) != std::string::npos) ||
```

Comment:
I have 2 opinions.
1. We have to add null check regarding mContentType.
2. This can be used another place in future, I think mediaUtils is better. (ex. getAudioTypeFromMimeType())

Reply:
Fixed, I added `utils::getAudioTypeFromMimeType(std::string &mimeType)`